### DISCUSSION TITLE
Residual loss always reach 1 and stops

### DISCUSSION THREAD

**Discussion by Axeor1997** [jax-ml/jax#17855] [Q&A]
Hey I try to training a Physics Informed Deep Operator Network and this is my model:
```python
# Define the model
class PI_DeepONet:
    def __init__(self, branch_layers, trunk_layers):    
        # Network initialization and evaluation functions
        self.branch_init, self.branch_apply = modified_deeponet(branch_layers, activation=swish)
        self.trunk_init, self.trunk_apply = modified_deeponet(trunk_layers, activation=swish)

        # Initialize
        branch_params = self.branch_init(rng_key = random.PRNGKey(1234))
        trunk_params = self.trunk_init(rng_key = random.PRNGKey(4321))
        params = (branch_params, trunk_params)

        # Use optimizers to set optimizer initialization and update functions
        self.opt_init, \
        self.opt_update, \
        self.get_params = optimizers.adam(optimizers.exponential_decay(1e-3, 
                                                                      decay_steps=5000, 
                                                                      decay_rate=0.9))
        self.opt_state = self.opt_init(params)

        _, self.unravel = ravel_pytree(params)

        # Logger
        self.itercount = itertools.count()
        self.loss_log = []
        self.loss_ics_log = []
        self.loss_res_log = []

    # Define DeepONet architecture
    def operator_net(self, params, u, y):
        branch_params, trunk_params = params
        B = self.branch_apply(branch_params, u)
        T = self.trunk_apply(trunk_params, y)

        B1 = B[:100]
        B2 = B[100:200]
        B3 = B[200:300]
        B4 = B[300:]

        T1 = T[:100]
        T2 = T[100:200]
        T3 = T[200:300]
        T4 = T[300:]

        s1 = np.sum(B1 * T1)
        s2 = np.sum(B2 * T2)
        s3 = np.sum(B3 * T3)
        s4 = np.sum(B4 * T4)

        #rescale s1 , s2
        return s1, s2, s3, s4

    # Define ODE residual
    def residual_net(self, params, u, y):

        s1, s2, s3, s4 = self.operator_net(params, u, y)

        def s1_fn(params, u, y):
          s1_fn, _ , _ , _ = self.operator_net(params, u, y)
          return s1_fn
        
        def s2_fn(params, u, y):
          _ , s2_fn, _ , _  = self.operator_net(params, u, y)
          return s2_fn

        def s3_fn(params, u, y):
            _ , _ , s3_fn, _ = self.operator_net(params, u, y)
            return s3_fn

        def s4_fn(params, u, y):
            _ , _ , _ , s4_fn = self.operator_net(params, u, y)
            return s4_fn

        

        s1_y = grad(s1_fn, argnums= 2)(params, u, y)
        s2_y = grad(s2_fn, argnums= 2)(params, u, y)
        s3_y = grad(s3_fn, argnums= 2)(params, u, y)
        s4_y = grad(s4_fn, argnums= 2)(params, u, y)

        res_1 = s1_y - s2
        res_2 = s2_y + (s3/680*4)
        res_3 = s3_y - s4
        res_4  = s4_y + (-1)

        return res_1, res_2, res_3, res_4

    # Define initial loss
    def loss_ics(self, params, batch):
        # Fetch data
        inputs, outputs = batch
        u, y = inputs

        s1 = outputs[:, 0:1]
        s2 = outputs[:, 1:2]
        s3 = outputs[:, 2:3]
        s4 = outputs[:, 3:4]

        # Compute forward pass
        s1_pred, s2_pred, s3_pred, s4_pred = vmap(self.operator_net, (None, 0, 0))(params, u, y)
        # Compute loss
        loss_s1 = np.mean((s1.flatten() - s1_pred)**2)
        loss_s2 = np.mean((s2.flatten() - s2_pred)**2)
        loss_s3 = np.mean((s3.flatten() - s3_pred)**2)
        loss_s4 = np.mean((s4.flatten() - s4_pred)**2)

        loss = loss_s1 + loss_s2 + loss_s3 + loss_s4
        return loss

    # Define residual loss
    def loss_res(self, params, batch):
        # Fetch data
        inputs, outputs = batch
        u, y = inputs

        # Compute forward pass
        res1_pred, res2_pred, res3_pred, res4_pred = vmap(self.residual_net, (None, 0, 0))(params, u, y)
        # Compute loss
        loss_res1 = np.mean((res1_pred)**2)
        loss_res2 = np.mean((res2_pred)**2)
        loss_res3 = np.mean((res3_pred)**2)
        loss_res4 = np.mean((res4_pred)**2)
        loss_res = loss_res1 + loss_res2 + loss_res3 + loss_res4
        return loss_res   

    # Define total loss
    def loss(self, params, ics_batch, res_batch):
        loss_ics = self.loss_ics(params, ics_batch)
        loss_res = self.loss_res(params, res_batch)
        loss =  loss_ics + loss_res
        return loss

    # Define a compiled update step
    @partial(jit, static_argnums=(0,))
    def step(self, i, opt_state, ics_batch, res_batch):
        params = self.get_params(opt_state)
        g = grad(self.loss)(params, ics_batch, res_batch)
        return self.opt_update(i, g, opt_state)

    # Optimize parameters in a loop
    def train(self, ics_dataset, res_dataset, nIter = 10000):
        ics_data = iter(ics_dataset)
        res_data = iter(res_dataset)

        pbar = trange(nIter)
        # Main training loop
        for it in pbar:
            ics_batch= next(ics_data)
            res_batch = next(res_data)

            self.opt_state = self.step(next(self.itercount), self.opt_state, ics_batch, res_batch)
            
            if it % 100 == 0:
                params = self.get_params(self.opt_state)

                # Compute losses
                loss_value = self.loss(params, ics_batch, res_batch)
                loss_ics_value = self.loss_ics(params, ics_batch)
                loss_res_value = self.loss_res(params, res_batch)

                # Store losses
                self.loss_log.append(loss_value)
                self.loss_ics_log.append(loss_ics_value)
                self.loss_res_log.append(loss_res_value)

                # Print losses
                pbar.set_postfix({'Loss': loss_value, 
                                  'loss_ics': loss_ics_value, 
                                  'loss_res': loss_res_value})
       
           
    # Evaluates predictions at test points  
    @partial(jit, static_argnums=(0,))
    def predict_s(self, params, U_star, Y_star):
        s1_pred, s2_pred, s3_pred, s4_pred = vmap(self.operator_net, (None, 0, 0))(params, U_star, Y_star)
        return s1_pred, s2_pred, s3_pred, s4_pred
```

Every things seems fine but my loss_res stuck on 1.0 while my loss_ic works...
any suggestion? i really need to fix this
